genesyscloud/resource_genesyscloud_tf_export.go

```diff
@@ -0,0 +1,594 @@
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"directory": {
+				Description: "Directory where the config and state files will be exported.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "./genesyscloud",
+				ForceNew:    true,
+			},
+			"resource_types": {
+				Description: "Resource types to export, e.g. 'genesyscloud_user'. Defaults to all exportable types.",
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validation.StringInSlice(getAvailableExporterTypes(), false),
+				},
+				ForceNew: true,
+			},
+			"include_state_file": {
+				Description: "Export a 'terraform.tfstate' file along with the config file. This can be used for orgs to begin managing existing resources with terraform.",
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				ForceNew:    true,
+			},
+		},
+	}
+}
+
+type resourceInfo struct {
+	State   *terraform.InstanceState
+	Name    string
+	Type    string
+	CtyType cty.Type
+}
+
+func createTfExport(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	jsonFilePath, diagErr := getFilePath(d, defaultTfJSONFile)
+	if diagErr != nil {
+		return diagErr
+	}
+
+	version := meta.(*providerMeta).Version
+	exporters := getResourceExporters()
+
+	// Apply resource type filter
+	if types, ok := d.GetOk("resource_types"); ok {
+		allowedTypes := *setToStringList(types.(*schema.Set))
+		for resType := range exporters {
+			if !stringInSlice(resType, allowedTypes) {
+				delete(exporters, resType)
+			}
+		}
+	}
+
+	if len(exporters) == 0 {
+		return diag.Errorf("No valid resource types to export.")
+	}
+
+	diagErr = buildSanitizedResourceMaps(exporters)
+	if diagErr != nil {
+		return diagErr
+	}
+
+	includeStateFile := d.Get("include_state_file").(bool)
+	provider := New(version)()
+
+	// Read the instance data from each exporter
+	var resources []resourceInfo
+	for resType, exporter := range exporters {
+		typeResources, err := getResourcesForType(resType, provider, exporter, meta)
+		if err != nil {
+			return err
+		}
+		resources = append(resources, typeResources...)
+	}
+
+	// Generate the JSON config map
+	resourceTypeJSONMaps := make(map[string]map[string]jsonMap)
+	for _, resource := range resources {
+		jsonResult, diagErr := instanceStateToJSONMap(resource.State, resource.CtyType)
+		if diagErr != nil {
+			return diagErr
+		}
+
+		// Removes zero values and sets proper reference expressions
+		sanitizeConfigMap(resource.Type, jsonResult, "", exporters, includeStateFile)
+
+		if resourceTypeJSONMaps[resource.Type] == nil {
+			resourceTypeJSONMaps[resource.Type] = make(map[string]jsonMap)
+		}
+		resourceTypeJSONMaps[resource.Type][resource.Name] = jsonResult
+	}
+
+	providerSource := "registry.terraform.io/mypurecloud/genesyscloud"
```

Comment:
I'd move this stuff off into another function e.g. `func sourceForVersion(version string) string`. Doesn't feel right to have this in the middle of this function.